### PR TITLE
fix race condition in throttle_timer; add service.Reset()

### PR DESCRIPTION
**Pull request by ebuchman**

### CODE REVIEW

File: throttle_timer.go [hunk 1]
```diff
@@ -36,19 +40,25 @@ func (t *ThrottleTimer) fireRoutine() {
...
 		// do nothing
 	default:
+		t.mtx.Lock()
 		t.timer.Reset(t.dur)
+		t.mtx.Unlock()
 	}
 }
 
 func (t *ThrottleTimer) Set() {
 	if atomic.CompareAndSwapUint32(&t.isSet, 0, 1) {
```
Comment (ebuchman) [inline throttle_timer.go]: since we need a mutex anyways, should we just drop this and use a bool

Comment (jaekwon) [inline throttle_timer.go]: ok

File: service.go [hunk 2]
```diff
@@ -120,6 +124,29 @@ func (bs *BaseService) Stop() bool {
 func (bs *BaseService) OnStop() {}
 
 // Implements Service
+func (bs *BaseService) Reset() (bool, error) {
+	if atomic.CompareAndSwapUint32(&bs.stopped, 1, 0) {
+		// whether or not we've started, we can reset
+		atomic.CompareAndSwapUint32(&bs.started, 1, 0)
+
+		return true, bs.impl.OnReset()
```
Comment (jaekwon) [inline service.go]: I think there's a race condition where someone can call Start() right after calling Reset() (on different goroutines), and cause Start() to get called before OnReset has had a chance to complete.  We should always run OnReset to completion before allowing Start to happen.

I think the easiest way would be to introduce another state for bs.started.  That is, if bs.started is 2, then it's in the process of restarting.

```
if atomic.LoadUint32(&bs.stopped) == 1 {
  if atomic.CompareAndSwapUint32(&bs.started, 1, 2) {
    err := bs.impl.OnReset()
    ...
    atomic.StoreUint32(&bs.stopped, 0)
    atomic.StoreUint32(&bs.started, 0)
  } else {
    return false, nil
  }
}
```

or something

Comment (ebuchman) [inline service.go]: yup, but there's also this warning, which should also apply to Reset: https://github.com/tendermint/go-common/blob/master/service.go#L7

**Comment by ebuchman** [comment]
throttle fix in https://github.com/tendermint/go-common/pull/4
closing until we need to revisit the restart thing

**Pull request closed without merging**